File: Leet Code Solutions/0341-flatten-nested-list-iterator/0341-flatten-nested-list-iterator.java
public class NestedIterator implements Iterator<Integer> {
    // The list of NestedInteger elements to be flattened
    private List<NestedInteger> nestedList;
    
    // The flattened list of integers
    private List<Integer> flattenedList;
    
    // Index to keep track of the current position in the flattenedList
    private int currentIndex = 0;

    public NestedIterator(List<NestedInteger> nestedList) {
        this.nestedList = nestedList;
        this.flattenedList = new ArrayList<>();
        // Flatten the nestedList during initialization
        this.flatten(nestedList);
    }

    // Returns the next integer in the flattened list
    @Override
    public Integer next() {
        int number = this.flattenedList.get(currentIndex);
        currentIndex++;
        return number;
    }

    // Checks if there are more integers in the flattened list
    @Override
    public boolean hasNext() {
        return currentIndex < flattenedList.size();
    }

    // Recursively flattens the nested list and adds integers to the flattenedList
    private void flatten(List<NestedInteger> currentList) {
        for (int i = 0; i < currentList.size(); i++) {
            if (currentList.get(i).isInteger()) {
                flattenedList.add(currentList.get(i).getInteger());
            } else {
                // Recursively flatten nested lists
                flatten(currentList.get(i).getList());
            }
        }
    }
}

File: Leet Code Solutions/0341-flatten-nested-list-iterator/0341-flatten-nested-list-iterator.cpp
class NestedIterator {
private:
    // The list of NestedInteger elements to be flattened
    vector<NestedInteger> nestedList;

    // The flattened list of integers
    vector<int> flattenedList;

    // Index to keep track of the current position in the flattenedList
    int currentIndex = 0;

    // Recursively flattens the nested list and adds integers to the flattenedList
    void flatten(vector<NestedInteger>& currentList) {
        for (int i = 0; i < currentList.size(); i++) {
            if (currentList[i].isInteger()) {
                flattenedList.push_back(currentList[i].getInteger());
            } else {
                // Recursively flatten nested lists
                flatten(currentList[i].getList());
            }
        }
    }

public:
    NestedIterator(vector<NestedInteger> &nestedList) {
        this->nestedList = nestedList;
        // Flatten the nestedList during initialization
        flatten(nestedList);
    }

    // Returns the next integer in the flattened list
    int next() {
        int number = flattenedList[currentIndex];
        currentIndex++;
        return number;
    }

    // Checks if there are more integers in the flattened list
    bool hasNext() {
        return currentIndex < flattenedList.size();
    }
};